File: web/lib/dropbox-client.ts
/**
 * lib/dropbox-client.ts — Dropbox REST API operations
 *
 * Handles folder creation, sharing links, folder moves, and subfolder listing.
 * Uses the Dropbox HTTP API directly (no Python SDK needed).
 * Authentication is handled by lib/dropbox-auth.ts.
 *
 * Depends on: lib/dropbox-auth.ts
 * Used by: lib/core.ts, lib/auto-creator.ts, lib/folder-mover.ts
 */

import { getDropboxToken } from "./dropbox-auth";

const API = "https://api.dropboxapi.com/2";

// Helper: POST to a Dropbox API endpoint with a JSON body
async function post(endpoint: string, body: unknown): Promise<unknown> {
  const token = await getDropboxToken();
  const res = await fetch(`${API}${endpoint}`, {
    method: "POST",
    headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });
  // 409 = conflict (folder already exists), handled by callers
  if (!res.ok && res.status !== 409) {
    const text = await res.text();
    throw new Error(`Dropbox API error ${res.status} on ${endpoint}: ${text}`);
  }
  return res.json();
}

/**
 * Create a folder at the given Dropbox path.
 * Returns true if created, false if it already existed.
 * path — Full Dropbox path, e.g. "/Creative 2026/Marketing Ads/My Task"
 */
export async function createFolder(path: string): Promise<boolean> {
  const res = await fetch(`${API}/files/create_folder_v2`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${await getDropboxToken()}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ path, autorename: false }),
  });
  if (res.status === 409) return false; // Already exists
  if (!res.ok) {
    const text = await res.text();
    throw new Error(`Dropbox create_folder failed ${res.status}: ${text}`);
  }
  return true;
}

/**
 * Get or create a shareable Dropbox link for the folder at path.
 * Returns the URL string.
 */
export async function getSharedLink(path: string): Promise<string> {
  const token = await getDropboxToken();
  const headers = { Authorization: `Bearer ${token}`, "Content-Type": "application/json" };

  // Try to create a new shared link
  const createRes = await fetch(`${API}/sharing/create_shared_link_with_settings`, {
    method: "POST", headers, body: JSON.stringify({ path }),
  });

  if (createRes.ok) {
    const data = await createRes.json() as { url: string };
    return data.url;
  }

  // If a link already exists, fetch and return it
  if (createRes.status === 409) {
    const listRes = await fetch(`${API}/sharing/list_shared_links`, {
      method: "POST", headers, body: JSON.stringify({ path, direct_only: true }),
    });
    const listData = await listRes.json() as { links: { url: string }[] };
    if (listData.links?.length) return listData.links[0].url;
  }

  throw new Error(`Could not get shared link for ${path}: ${createRes.status}`);
}

/**
 * Resolve a Dropbox shared link URL to its actual folder path.
 * Returns the path string (e.g. "/Creative 2026/...") or null if lookup fails.
 */
export async function getFolderPathFromLink(sharedLinkUrl: string): Promise<string | null> {
  try {
    const token = await getDropboxToken();
    const headers = { Authorization: `Bearer ${token}`, "Content-Type": "application/json" };
    const meta = await fetch(`${API}/sharing/get_shared_link_metadata`, {
      method: "POST", headers, body: JSON.stringify({ url: sharedLinkUrl }),
    });
    const metaData = await meta.json() as { path_lower?: string };
    if (!metaData.path_lower) return null;
    const file = await fetch(`${API}/files/get_metadata`, {
      method: "POST", headers, body: JSON.stringify({ path: metaData.path_lower }),
    });
    const fileData = await file.json() as { path_display?: string };
    return fileData.path_display ?? null;
  } catch {
    return null;
  }
}

/**
 * Move a Dropbox folder from one path to another.
 * Returns true on success, false if the move failed.
 */
export async function moveFolder(fromPath: string, toPath: string): Promise<boolean> {
  try {
    await post("/files/move_v2", { from_path: fromPath, to_path: toPath });
    return true;
  } catch {
    return false;
  }
}

/**
 * List the names of all direct subfolders at the given Dropbox path.
 * Returns an empty array if the folder doesn't exist or has no subfolders.
 */
export async function listSubfolderNames(path: string): Promise<string[]> {
  try {
    const token = await getDropboxToken();
    const res = await fetch(`${API}/files/list_folder`, {
      method: "POST",
      headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json" },
      body: JSON.stringify({ path }),
    });
    if (!res.ok) return [];
    const data = await res.json() as { entries: { ".tag": string; name: string }[] };
    return data.entries
      .filter((e) => e[".tag"] === "folder")
      .map((e) => e.name);
  } catch {
    return [];
  }
}

File: dropbox_client.py
"""
dropbox_client.py — Dropbox API communication

This file handles all interactions with Dropbox:
creating folders, listing existing folders, generating shareable links,
and moving folders (used by the Folder Mover web page).

Depends on: .env (for DROPBOX_ACCESS_TOKEN)
Used by: folder_builder.py, main.py, web_folder_mover.py
"""

import os
import sys
import dropbox
from dropbox.exceptions import ApiError
from dropbox.files import CreateFolderError, FolderMetadata
from dropbox.sharing import CreateSharedLinkWithSettingsError
from dotenv import load_dotenv

# Load DROPBOX_ACCESS_TOKEN from the .env file
load_dotenv()


def _get_client() -> dropbox.Dropbox:
    """
    Create and return an authenticated Dropbox client.

    Prefers the new OAuth2 refresh-token method (DROPBOX_REFRESH_TOKEN +
    DROPBOX_APP_KEY + DROPBOX_APP_SECRET), which never expires.

    Falls back to a legacy DROPBOX_ACCESS_TOKEN if the refresh credentials
    are not yet configured. Run get_dropbox_token.py once to get a refresh token.
    """
    refresh_token = os.getenv("DROPBOX_REFRESH_TOKEN")
    app_key = os.getenv("DROPBOX_APP_KEY")
    app_secret = os.getenv("DROPBOX_APP_SECRET")

    # Preferred: use refresh token so auth never expires
    if refresh_token and app_key and app_secret:
        return dropbox.Dropbox(
            oauth2_refresh_token=refresh_token,
            app_key=app_key,
            app_secret=app_secret,
        )

    # Fallback: legacy short-lived access token (expires every 4 hours)
    token = os.getenv("DROPBOX_ACCESS_TOKEN")
    if token:
        print(
            "  ⚠ Using DROPBOX_ACCESS_TOKEN which expires every 4 hours. "
            "Run get_dropbox_token.py to set up a permanent refresh token.",
            file=sys.stderr,
        )
        return dropbox.Dropbox(token)

    raise EnvironmentError(
        "No Dropbox credentials found. Add DROPBOX_REFRESH_TOKEN + DROPBOX_APP_KEY + "
        "DROPBOX_APP_SECRET to .env, or run get_dropbox_token.py."
    )


def create_folder(path: str) -> bool:
    """
    Create a folder at the given Dropbox path (including any missing parent folders).
    Returns True if the folder was created, False if it already existed.

    path — Full Dropbox path, e.g. '/Creative Production/Marketing/Face Cream/Video/Meta/001_March_2026/My Task'
    """
    dbx = _get_client()
    try:
        dbx.files_create_folder_v2(path)
        print(f"  ✓ Created folder: {path}", file=sys.stderr)
        return True
    except ApiError as e:
        # If the folder already exists, that's fine — not an error
        if isinstance(e.error, CreateFolderError) and e.error.is_path() and e.error.get_path().is_conflict():
            print(f"  ℹ Folder already exists: {path}", file=sys.stderr)
            return False
        raise  # Re-raise any other unexpected errors


def get_shared_link(path: str) -> str:
    """
    Get a shareable Dropbox link for the folder at the given path.
    If a link already exists, it returns the existing one instead of creating a duplicate.

    path — Full Dropbox path of the folder
    Returns the shareable URL as a string.
    """
    dbx = _get_client()
    try:
        # Try to create a new shared link
        result = dbx.sharing_create_shared_link_with_settings(path)
        return result.url
    except ApiError as e:
        # If a shared link already exists for this path, fetch and return it
        if isinstance(e.error, CreateSharedLinkWithSettingsError) and e.error.is_shared_link_already_exists():
            existing = dbx.sharing_list_shared_links(path=path, direct_only=True)
            if existing.links:
                return existing.links[0].url
        raise  # Re-raise anything else unexpected


def get_folder_path_from_link(shared_link_url: str) -> "str | None":
    """
    Get the actual Dropbox folder path from a shared link URL.
    Uses the Dropbox sharing API to look up metadata for the link.

    shared_link_url — A Dropbox shared link, e.g. 'https://www.dropbox.com/scl/fo/...'

    Returns the folder's display path (e.g. '/Creative 2026/Marketing Ads/...'),
    or None if the lookup fails.
    """
    dbx = _get_client()
    try:
        # FolderLinkMetadata has path_lower but not path_display
        link_meta = dbx.sharing_get_shared_link_metadata(shared_link_url)
        path_lower = link_meta.path_lower
        # Use files_get_metadata to resolve the properly-cased display path
        file_meta = dbx.files_get_metadata(path_lower)
        return file_meta.path_display
    except Exception as e:
        print(f"  ✗ Could not get path from shared link: {e}", file=sys.stderr)
        return None


def move_folder(from_path: str, to_path: str) -> bool:
    """
    Move a Dropbox folder from one location to another.
    The folder and all its contents are preserved — only the location changes.

    from_path — Current full path, e.g. '/Creative 2026/Marketing Ads/.../Old Name'
    to_path   — Destination full path, e.g. '/Creative 2026/Marketing Ads/.../New Name'

    Returns True on success, False if the move failed.
    """
    dbx = _get_client()
    try:
        dbx.files_move_v2(from_path, to_path)
        print(f"  ✓ Moved folder: {from_path} → {to_path}", file=sys.stderr)
        return True
    except Exception as e:
        print(f"  ✗ Failed to move folder: {e}", file=sys.stderr)
        return False


def list_subfolder_names(path: str) -> list:
    """
    List the names of all direct subfolders inside a given Dropbox path.
    Returns an empty list if the folder doesn't exist yet or has no subfolders.

    path — Full Dropbox path to inspect
    Used by folder_builder.py to compute the next sequential date index.
    """
    dbx = _get_client()
    try:
        result = dbx.files_list_folder(path)
        # Return only folder names (not files)
        return [
            entry.name
            for entry in result.entries
            if isinstance(entry, FolderMetadata)
        ]
    except ApiError:
        # If the folder doesn't exist yet, return an empty list (no entries = index starts at 1)
        return []
